File: 05-Algorithms/code/python/seb_alg/prim.py
"""Mirrors `typescript/src/prim.ts` exactly."""

from __future__ import annotations

import math
from typing import Dict, List, Sequence

from .graph_common import build_weighted_adj_list


def prim(n: int, edges: Sequence[Sequence[int]]) -> Dict[str, object]:
    """Minimum spanning forest via Prim's algorithm: grows a tree from the
    lowest-id unvisited vertex, array-scan for the cheapest crossing edge
    each step. Ties broken by lower "to" id, then lower "from" id. Restarts
    on unvisited vertices to cover disconnected graphs. O(V^2 + E).
    """
    adj = build_weighted_adj_list(n, edges, False)
    in_tree = [False] * n
    chosen: List[List[int]] = []
    weight = 0

    for seed in range(n):
        if in_tree[seed]:
            continue
        in_tree[seed] = True
        for _ in range(n):
            best_from = -1
            best_to = -1
            best_weight = math.inf
            for u in range(n):
                if not in_tree[u]:
                    continue
                for v, w in adj[u]:
                    if in_tree[v]:
                        continue
                    if w < best_weight or (
                        w == best_weight and (best_to == -1 or v < best_to or (v == best_to and u < best_from))
                    ):
                        best_weight = w
                        best_from = u
                        best_to = v
            if best_to == -1:
                break
            in_tree[best_to] = True
            chosen.append([best_from, best_to, best_weight])
            weight += best_weight

    return {"weight": weight, "edges": chosen}

File: 05-Algorithms/code/typescript/src/prim.ts
import { buildWeightedAdjList, type WeightedEdge } from "./graphCommon.js";
import { type MstResult } from "./kruskal.js";

/**
 * Minimum spanning forest via Prim's algorithm: grows a tree from the
 * lowest-id unvisited vertex, at each step adding the cheapest edge leaving
 * the current tree to an outside vertex (array-scan for the minimum, no
 * heap -- fine for this lab's graph sizes). Ties on edge weight are broken
 * by lower "to" vertex id, then lower "from" vertex id, keeping the result
 * deterministic. Restarts on the lowest-id unvisited vertex whenever the
 * current tree can't grow further, so disconnected graphs yield a full
 * minimum spanning forest (edges grouped by the order each component's
 * tree was grown, not sorted globally). O(V^2 + E).
 */
export function prim(n: number, edges: WeightedEdge[]): MstResult {
  const adj = buildWeightedAdjList(n, edges, false);
  const inTree = new Array(n).fill(false);
  const chosen: WeightedEdge[] = [];
  let weight = 0;

  for (let seed = 0; seed < n; seed++) {
    if (inTree[seed]) continue;
    inTree[seed] = true;
    for (let step = 0; step < n; step++) {
      let bestFrom = -1;
      let bestTo = -1;
      let bestWeight = Infinity;
      for (let u = 0; u < n; u++) {
        if (!inTree[u]) continue;
        for (const [v, w] of adj[u]) {
          if (inTree[v]) continue;
          if (w < bestWeight || (w === bestWeight && (bestTo === -1 || v < bestTo || (v === bestTo && u < bestFrom)))) {
            bestWeight = w;
            bestFrom = u;
            bestTo = v;
          }
        }
      }
      if (bestTo === -1) break;
      inTree[bestTo] = true;
      chosen.push([bestFrom, bestTo, bestWeight]);
      weight += bestWeight;
    }
  }

  return { weight, edges: chosen };
}
